src/d3a/models/state.py

```diff
@@ -43,131 +46,247 @@
-        self._desired_energy_Wh = {}
-        # Energy that the load needs to consume. It's reduced when new energy is bought
-        self._energy_requirement_Wh = {}
-        self._total_energy_demanded_Wh = 0
+class PVState(ProductionState):
+    """State class for PV devices.
 
-    def get_energy_requirement_Wh(self, time_slot, default_value=0.0):
-        if default_value is None:
-            return self._energy_requirement_Wh[time_slot]
-        return self._energy_requirement_Wh.get(time_slot, default_value)
+    Completely inherits ProductionState, but we keep this class for backward compatibility.
+    """
 
-    def get_desired_energy_Wh(self, time_slot, default_value=0.0):
-        if default_value is None:
-            return self._desired_energy_Wh[time_slot]
-        return self._desired_energy_Wh.get(time_slot, default_value)
+
+class LoadState(ConsumptionState):
+    def __init__(self):
+        super().__init__()
 
     @property
     def total_energy_demanded_Wh(self):
         return self._total_energy_demanded_Wh
 
-    def can_buy_more_energy(self, time_slot):
-        if time_slot not in self._energy_requirement_Wh:
-            return False
-        return self._energy_requirement_Wh[time_slot] > FLOATING_POINT_TOLERANCE
+    def get_desired_energy(self, time_slot):
+        return self._desired_energy_Wh[time_slot]
 
-    def calculate_energy_to_accept(self, offer_energy_Wh, time_slot):
-        if offer_energy_Wh > self._energy_requirement_Wh[time_slot]:
-            return self._energy_requirement_Wh[time_slot]
-        else:
-            return offer_energy_Wh
 
-    def decrement_energy_requirement(
-            self, purchased_energy_Wh: float, time_slot: DateTime, area_name: str) -> None:
-        """Decrease the energy required by the device in a specific market slot."""
-        self._energy_requirement_Wh[time_slot] -= purchased_energy_Wh
-        assert self._energy_requirement_Wh[time_slot] >= -FLOATING_POINT_TOLERANCE, \
-            f"Energy requirement for load {area_name} fell below zero " \
-            f"({self._energy_requirement_Wh[time_slot]})."
+class HomeMeterState(ConsumptionState, ProductionState):
+    """State for the Home Meter device"""
 
-    def set_desired_energy(self, energy, time_slot, overwrite=False):
-        if overwrite is False and time_slot in self._energy_requirement_Wh:
-            return
-        self._energy_requirement_Wh[time_slot] = energy
-        self._desired_energy_Wh[time_slot] = energy
+    def __init__(self):
+        super().__init__()
 
-    def get_desired_energy(self, time_slot):
-        return self._desired_energy_Wh[time_slot]
+    def _validate_state(self):
+        production_market_slots = {
+            slot for slot, energy in self._available_energy_kWh.items() if energy}
+        consumption_market_slots = {
+            slot for slot, energy in self._energy_requirement_Wh.items() if energy}
```

Comment:
Every time the `market_slot` property is called, this validation is performed. IMO an overkill. 
AFAIU, you are already checking here https://github.com/gridsingularity/d3a/pull/1159/files#diff-1187ba841fc3bdd8c4b0200dc3033dfc48bd59ba5c22485e605fd0d1293e1d78R438 If produced and consumed both have non-zero numbers. This is enough IMO. For the sake of performance, i suggest to remove this additional validation here and only  return the union like you do in the market_slots property. 
This is especially valid for high number if future markets.
Or did I miss some corner case?

Reply:
I think your concern is well-founded. This validation can become slow for cases where we have many markets, and in theory this situation shouldn't occur given the structure of the profiles. 
Just to be sure: you're suggesting to completely remove the `_validate_state()` method and not call it in the `market_slots` property, right?

Reply:
Yes